packages/algolia/src/client.ts

```diff
@@ -2,103 +2,103 @@ import { SearchOptions, SearchResponse } from '@algolia/client-search';
 import {
   EventResponse,
   ExternalAuthorResponse,
-  LabResponse,
+  gp2 as gp2Model,
   ResearchOutputResponse,
   UserResponse,
 } from '@asap-hub/model';
 import { SearchIndex } from 'algoliasearch';
+import {
+  EVENT_ENTITY_TYPE,
+  EXTERNAL_AUTHOR_ENTITY_TYPE,
+  Payload,
+  RESEARCH_OUTPUT_ENTITY_TYPE,
+  USER_ENTITY_TYPE,
+} from './crn';
+import { OUTPUT_ENTITY_TYPE, Payload as GP2Payload } from './gp2';
 
-export const RESEARCH_OUTPUT_ENTITY_TYPE = 'research-output';
-export const USER_ENTITY_TYPE = 'user';
-export const EXTERNAL_AUTHOR_ENTITY_TYPE = 'external-author';
-export const LAB_ENTITY_TYPE = 'lab';
-export const EVENT_ENTITY_TYPE = 'event';
-
-export type Payload =
-  | {
-      data: EventResponse;
-      type: 'event';
-    }
-  | {
-      data: ExternalAuthorResponse;
-      type: 'external-author';
-    }
-  | {
-      data: LabResponse;
-      type: 'lab';
-    }
-  | {
-      data: ResearchOutputResponse;
-      type: 'research-output';
-    }
-  | {
-      data: UserResponse;
-      type: 'user';
-    };
-
+const CRN = 'crn';
+const GP2 = 'gp2';
+export type Apps = typeof CRN | typeof GP2;
 export type EntityResponses = {
-  [RESEARCH_OUTPUT_ENTITY_TYPE]: ResearchOutputResponse;
-  [USER_ENTITY_TYPE]: UserResponse;
-  [EXTERNAL_AUTHOR_ENTITY_TYPE]: ExternalAuthorResponse;
-  [EVENT_ENTITY_TYPE]: EventResponse;
+  [CRN]: {
+    [RESEARCH_OUTPUT_ENTITY_TYPE]: ResearchOutputResponse;
+    [USER_ENTITY_TYPE]: UserResponse;
+    [EXTERNAL_AUTHOR_ENTITY_TYPE]: ExternalAuthorResponse;
+    [EVENT_ENTITY_TYPE]: EventResponse;
+  };
+  [GP2]: {
+    [OUTPUT_ENTITY_TYPE]: gp2Model.OutputResponse;
+  };
 };
-
-export type EntityRecord<T extends keyof EntityResponses> =
-  EntityResponses[T] & {
-    objectID: string;
-    __meta: {
-      type: T;
-    };
+type SavePayload = Payload | GP2Payload;
+type DistributeToEntityRecords<
+  Responses extends EntityResponses[Apps],
+  ResponsesKey extends keyof Responses,
+> = Responses[ResponsesKey] & {
+  objectID: string;
+  __meta: {
+    type: ResponsesKey;
   };
+};
+export type ClientSearchResponse<
+  App extends Apps,
+  ResponsesKey extends keyof EntityResponses[App],
+> = SearchResponse<
+  DistributeToEntityRecords<EntityResponses[App], ResponsesKey>
+>;
+export type ClientSearch<
```

Comment:
this type and its naming is a little hard to understand, from what I can tell you are defining a method separately and then applying that within the client interface. Have you considered having that definition only as part of the client interface and then referencing that particular method in your fixtures/etc? You can add a generic to the interface if thats something you need

Reply:
I wanted to do

```type ClientSearchResponse = ReturnType<SearchClient['search']>;```

But I was getting `Type ClientSearchResponse is not generic`